### PR TITLE
- added category plugin that enables adding category for the filters.

### PR DESCRIPTION
**Pull request by cansaner**
**Merge request checklist**

- [x] I read the [guidelines for contributing](https://github.com/mistic100/jQuery-QueryBuilder/blob/master/.github/CONTRIBUTING.md)
- [x] I created my branch from `dev` and I am issuing the PR to `dev`
- [ ] Unit tests are OK
- [ ] If it's a new feature, I added the necessary unit tests
- [ ] If it's a new language, I filled the `__locale` and `__author` fields

### CODE REVIEW

File: src/plugins/category/plugin.js [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+     */
+    this.on('ruleToJson.filter', function(e, rule) {
+        e.value.category = rule.category.id;
+    });
+
+    /**
+     * Read "category" from JSON
+     */
+    this.on('jsonToRule.filter', function(e, json) {
+        //e.value.category = e.builder.getCategoryById(json.category ? json.category : '-1');
```
Comment (mistic100) [inline src/plugins/category/plugin.js]: need to be implemented

File: examples/index.html [hunk 2]
```diff
@@ -16,9 +16,7 @@
 
   <link rel="stylesheet" href="http://mistic100.github.io/jQuery-QueryBuilder/assets/flags/flags.css">
   <style>
-    .flag {
```
Comment (mistic100) [inline examples/index.html]: please do not change the indentation of the whole file.

File: src/plugins/category/plugin.js [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+});
+
+QueryBuilder.define('category', function(options) {
+    var self = this;
+
+    // SETTINGS SHORTCUT
+    this.categories = this.settings.categories;
+
+    // CHECK ALL CATEGORIES
+    this.categories = this.checkCategories(this.categories);
```
Comment (mistic100) [inline src/plugins/category/plugin.js]: merge the two lines

File: src/plugins/category/plugin.js [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+    container: 'body',
+    style: 'btn-inverse btn-xs',
+    width: 'auto',
+    showIcon: false
+});
+
+// keep a pointer to the original addRule method
+var originaAddRule = QueryBuilder.prototype.addRule;
+
+QueryBuilder.extend({
```
Comment (mistic100) [inline src/plugins/category/plugin.js]: Not ok, you must find a way the call the original method (but putting the new code before or after) or use existing events, or add new events where needed.

Comment (mistic100) [inline src/plugins/category/plugin.js]: Same for all overwritten methods

File: src/plugins/category/plugin.js [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+    /**
+     * Decide if filter is available for category
+     * @param filterId {string}
+     * @param categoryId {string}
+     * @return {boolean} true if filter belongs to that category
+     */
+    isFilterOKForCategory: function(filterId, categoryId) {
+        var filtersOfCategory = this.getFiltersForCategoryId(categoryId);
+
+        for (var i = 0; i < filtersOfCategory.length; i++) {
```
Comment (mistic100) [inline src/plugins/category/plugin.js]: can make usage of `filtersOfCategory.some` method

File: src/plugins/category/plugin.js [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+
+    /**
+     * Filter filters array according to categories of filters 
+     * @param categoryId {string}
+     * @return {array}
+     */
+    getFiltersForCategoryId: function(categoryId) {
+        var filtersOfCategory = [];
+        var category = this.getCategoryById(categoryId);
+        this.filters.forEach(function(filter, i) {
```
Comment (mistic100) [inline src/plugins/category/plugin.js]: useless double foreach : loop over `category.filters` and use `getFilterById`

File: src/plugins/category/plugin.js [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+    },
+
+    /**
+     * Decide if filter is available for category
+     * @param filterId {string}
+     * @param categoryId {string}
+     * @return {boolean} true if filter belongs to that category
+     */
+    isFilterOKForCategory: function(filterId, categoryId) {
+        var filtersOfCategory = this.getFiltersForCategoryId(categoryId);
```
Comment (mistic100) [inline src/plugins/category/plugin.js]: useless access to `getFiltersForCategoryId`: use `getCategoryById` and loop over its `filters`

**Comment by mistic100** [comment]
The CI failed on codestyle, run tests locally before pushing.

**Comment by mistic100** [comment]
@cansaner there are two much breakable code (overriden methods) for this to be merged.

You are free to release this as a separated standalone plugin in it's own repository (I can help you to make it standalone if you want).

**Comment by cansaner** [comment]
@mistic100 I will definitely focus on making it standalone. My thesis statement is pushing me hard these days. I ll hopefully focus back on October I guess... Thanks a lot in advance

**Pull request closed without merging**